File: katas/4 kyu/twice-linear/solution.py
# twice-linear
import codewars_test as test
# -------------------------------- MY SOLUTION ------------------------------- #
def dbl_linear(n) : 
  list = [1] 
  x = 0
  y = 0
  
  while(len(list)<=n) : 
    a = 2*list[x] + 1 
    b = 3*list[y] + 1 
    
    if a>b : 
      list.append(b)
      y+= 1 
    elif a<b : 
      list.append(a)
      x += 1 
    else : 
      list.append(a)
      x += 1 
      y += 1

  return list[n]
# ------------------------------ CLEVER SOLUTION ----------------------------- #
# ----------------------------------- TEST ----------------------------------- #
test.assert_equals(dbl_linear(10), 22)
test.assert_equals(dbl_linear(20), 57)
test.assert_equals(dbl_linear(30), 91)
test.assert_equals(dbl_linear(50), 175)

File: katas/4 kyu/twice-linear/solution.php
<!-- twice-linear -->
<?php
use PHPUnit\Framework\TestCase;
/* ------------------------------- MY SOLUTION ------------------------------ */
function dblLinear($n) {
  $u = [1];
  $i = 0;
  $j = 0;

  while (count($u) <= $n) {
    $y = 2 * $u[$i] + 1;
    $z = 3 * $u[$j] + 1;

    if ($y > $z) {
      array_push($u , $z);
      $j++;
    }

    elseif ($y < $z) {
      array_push($u , $y);
      $i++;
    }

    else {
      array_push($u , $y);
      $i++;
      $j++;
    }
  }
  return $u[$n];
}
/* ----------------------------- CLEVER SOLUTION ---------------------------- */
/* ---------------------------------- TEST ---------------------------------- */
class DoubleLinearTest extends TestCase {
  private function revTest($actual, $expected) {
      $this->assertSame($expected, $actual);
  }
  public function testBasics() {        
      $this->revTest(dblLinear(10), 22);
      // $this->revTest(dblLinear(20), 57);
      $this->revTest(dblLinear(30), 91);
      $this->revTest(dblLinear(50), 175);
      $this->revTest(dblLinear(100), 447);
  }
}